zebra-rpc/src/queue.rs

```diff
@@ -0,0 +1,323 @@
+                let _ = &self.queue.insert(tx.clone());
+            }
+
+            if !self.queue.transactions().is_empty() {
+                // remove what is expired
+                self.remove_expired(spacing);
+
+                // remove if any of the queued transactions is now in the mempool
+                let in_mempool =
+                    Self::check_mempool(mempool.clone(), self.transactions_as_hash_set()).await;
+                self.remove_committed(in_mempool);
+
+                // remove if any of the queued transactions is now in the state
+                let in_state =
+                    Self::check_state(state.clone(), self.transactions_as_hash_set()).await;
+                self.remove_committed(in_state);
+
+                // retry what is left in the queue
+                let _retried = Self::retry(mempool.clone(), self.transactions_as_vec()).await;
+            }
+        }
+    }
+
+    /// Remove transactions that are expired according to number of blocks and current spacing between blocks.
+    fn remove_expired(&mut self, spacing: Duration) {
+        // Have some extra time to to make sure we re-submit each transaction `NUMBER_OF_BLOCKS_TO_EXPIRE`
+        // times, as the main loop also takes some time to run.
+        let extra_time = Duration::seconds(5);
+
+        let duration_to_expire =
+            Duration::seconds(NUMBER_OF_BLOCKS_TO_EXPIRE * spacing.num_seconds()) + extra_time;
+        let transactions = self.queue.transactions();
+        let now = Instant::now();
+
+        for tx in transactions.iter() {
+            let tx_time =
+                tx.1 .1
+                    .checked_add(
+                        duration_to_expire
+                            .to_std()
+                            .expect("should never be less than zero"),
+                    )
+                    .expect("this is low numbers, should always be inside bounds");
+
+            if now > tx_time {
+                self.queue.remove(*tx.0);
+            }
+        }
+    }
+
+    /// Remove transactions from the queue that had been inserted to the state or the mempool.
+    fn remove_committed(&mut self, to_remove: HashSet<UnminedTxId>) {
+        for r in to_remove {
+            self.queue.remove(r);
+        }
+    }
+
+    /// Check the mempool for given transactions.
+    ///
+    /// Returns transactions that are in the mempool.
+    async fn check_mempool<Mempool>(
+        mempool: Mempool,
+        transactions: HashSet<UnminedTxId>,
+    ) -> HashSet<UnminedTxId>
+    where
+        Mempool: Service<Request, Response = Response, Error = BoxError> + Clone + 'static,
+    {
+        let mut response = HashSet::new();
+
+        if !transactions.is_empty() {
+            let request = Request::TransactionsById(transactions);
+
+            let mempool_response = mempool
+                .oneshot(request)
+                .await
+                .expect("Requesting transactions should not panic");
```

Comment:
Question: what if we change the mempool and it starts returning errors?

```suggestion
                .expect("Requesting transactions should not error");
```

Reply:
Let's ignore any errors returned by the mempool.

Reply:
Done as part of https://github.com/ZcashFoundation/zebra/pull/4015/commits/f82528104bd9b0c8573f196958f61cdd43a6652b